rosjava/src/main/java/org/ros/internal/node/server/SlaveServer.java

```diff
@@ -81,13 +86,22 @@ public void start() {
     super.start(org.ros.internal.node.xmlrpc.SlaveXmlRpcEndpointImpl.class,
         new SlaveXmlRpcEndpointImpl(this));
     tcpRosServer.start();
+    shutdownStarted = false;
   }
 
   // TODO(damonkohler): This should also shut down the Node.
   @Override
   public void shutdown() {
+    // prevent recursive call of this method
+    if (shutdownStarted) {
+      return;
+    }
+    shutdownStarted = true;
```

Comment:
About thread safety:
if `shutdown` is called a second time after the if statement is processed but before `shutdownStarted` is set to `true`, it will be called twice.
To be completely sure, I'd suggest placing all the code in both `shutdown` and `start` in a `synchronized` block using a dummy object as the lock. E.g.:

```java
  // In the constructor:
  shutdownLock = new Object();

  // Then:
  @Override
  public void shutdown() {
      synchronized(shutdownLock) {
          // current code goes here
      }
  }
```

And the same for `start`.